# Variational Calculation: Linear Combination Potential Well

**Summary:** Consider the ground state wavefunction $\psi_0$ and energy $E_0$ that satisfy the Schr$\ddot{o}$dinger equation: 

$$\hat{H}\psi_0=E_0\psi_0$$

The variational principle states that if any other wavefuction $\phi$ is used, then the resulting energy will be greater than the ground state energy

$$E_\phi \geq E_0$$

This principle is demonstrated on the infinitely deep potential well.

**Background:** Analytical solutions for the wavefunction of a system cannot be obtained for any but the most simple systems such as the hydrogen atom and the infinitely deep potential well. If the ground state wavefunction is not known, a trial function $\phi$ can be used in its place. According to the variational principle, $\psi_0$ represents a lower bound on energy $E_0$ and any trial function $\phi$ will result in an energy of at least $E_0$. Therefore, the trial function that results in the lowest possible energy could serve as an approximation for $\psi_0$.

Recall that:

$$E(\phi)=\frac{\int \phi^* H \phi}{\int \phi^* \phi}$$

And that if $\phi$ is a linear combination of primitive functions $\chi_m$, then the problem can be reduced to an eigenvalue problem 

$$\mathbf{HC}=E\mathbf{SC}$$

where,<br><br> 

$$\mathbf{H}_{mn} = <\chi_m|H|\chi_n>\;\;\;\;\;\;\;\;\mathbf{S} = <\chi_m|\chi_n>$$

**Setup:** For the infinitely deep potential well, we will consider trial functions of the type:

$$\phi_n(x)=\sum_0^n x^n(x-1)(x+1)$$

$\phi_n$ is a linear combination of primitive functions $\chi_m$: $(x-1)(x+1),\;x(x-1)(x+1),\;x^2(x-1)(x+1)$, etc.

With these primitive functions the matrix elements to the overlap **S** and Hamilton **H** matrix can calculated analytically and are given as:

$$\mathbf{S}_{mn} = \frac{2}{m+n+5} - \frac{4}{m+n+3} + \frac{2}{m+n+1},\;\;for\;m\;+\;n\;even;\;otherwise\;\mathbf{S}_{mn}=0$$

$$\mathbf{H}_{mn} = -8\frac{1-m-n-2mn}{(m+n+3)(m+n+1)(m+n-1)},\;\;for\;m\;+\;n\;even;\;otherwise\;\mathbf{H}_{mn}=0$$

**Code Implementation:** Given the matrices, Python can easily compute the eigenvalues and eigenvectors with its linear algebra module. Let's import that module here:

In [6]:
import scipy.linalg as ln
import numpy as np

Now we will define functions that produce the correct value for the $\mathbf{S}$ and $\mathbf{H}$ matrix elements given an index $m,\;n$.  

In [7]:
def inf_well_S(m, n):
    if (n + m + 1) % 2:
        t1 = 2 / (n + m + 5)
        t2 = -4 / (n + m + 3)
        t3 = 2 / (n + m + 1)
        return t1 + t2 + t3
    else:
        return 0


def inf_well_H(m, n):
    if (n + m + 1) % 2:
        num = 1 - m - n - 2 * m * n
        den = (m + n + 3) * (m + n + 1) * (m + n - 1)
        return -8 * num / den
    else:
        return 0

Next, given $n$ number of primitive functions we need to create our $\mathbf{S}$ and $\mathbf{H}$ matrices and populate them. Let's create a function to do this.

In [8]:
def pop_matrices(n):
    H = np.zeros((n, n))
    S = np.zeros((n, n))
    for i in range(n):
        for j in range(n):
            H[i][j] = inf_well_H(i, j)
            S[i][j] = inf_well_S(i, j)
    return H, S

We need a function to evaluate the eigenvalues. We'll call Python's linear algebra module which we previously imported.

In [9]:
def compute_E(n):
    return ln.eigvals(*pop_matrices(n))

Let's run our code using eight different trial functions $\phi$. These trial functions will differ by the number of primitive functions that we use. We will use 1, 2, 3, 4, 5, 8, 12, and 16 primitive functions in our eight trial functions, respectively. 

In [15]:
result = list()
counter = 0
for i in [1, 2, 3, 4, 5, 8, 12, 16]:
    result.append([i])
    result[counter].append(np.sort(ln.eigvals(*pop_matrices(i)))[:5].real)
    counter += 1

for i in result:
    print('Number of primitve functions used: {}'.format(i[0]))
    print('Lowest five (or less) eigenvalues: {}'.format(i[1]))
    print('')

Number of primitve functions used: 1
Lowest five (or less) eigenvalues: [ 2.5]

Number of primitve functions used: 2
Lowest five (or less) eigenvalues: [  2.5  10.5]

Number of primitve functions used: 3
Lowest five (or less) eigenvalues: [  2.46743741  10.5         25.53256259]

Number of primitve functions used: 4
Lowest five (or less) eigenvalues: [  2.46743741   9.8753882   25.53256259  50.1246118 ]

Number of primitve functions used: 5
Lowest five (or less) eigenvalues: [  2.46740111   9.8753882   22.29340591  50.1246118   87.73919298]

Number of primitve functions used: 8
Lowest five (or less) eigenvalues: [  2.4674011    9.86960441  22.20736704  39.48924103  63.60446044]

Number of primitve functions used: 12
Lowest five (or less) eigenvalues: [  2.4674011    9.8696044   22.20660991  39.47841793  61.68624765]

Number of primitve functions used: 16
Lowest five (or less) eigenvalues: [  2.4674011    9.8696044   22.2066099   39.4784176   61.68502755]



The lowest eigenvalue represents the ground state energy. The exact (analytical) ground state energy is 2.4674011. As can be seen, as the number of primitive functions increases, the ground state energy is approximated with high accuracy. 

Can any physical meaning be assigned to the other eigenvalues? Yes, and no...

The exact energy for the first four levels above the groundstate are: 9.8696, 22.2066, 38.4784, and 61.6850.
So it would appear that as the number of primitive functions increases, the energy levels near the ground state are also approximated to a better degree. While this is true, the infinitely deep potential well is still a relatively simple problem, and in practice, the higher energy eigenvalues will not satisfactorily approximate the higher energy levels. That is, the variational principle really only guarantees that the ground state energy will be accurately approximated (to the degree that the primitive functions can describe the wavefunction). Note that other methods exist to accurately approximate the upper bound on the energy.